internal/unionstore/art/art_node_test.go

```diff
@@ -0,0 +1,259 @@
+	}
+
+	addr, n4 := allocator.allocNode4()
+	testFn(&artNode{kind: typeNode4, addr: addr}, &n4.nodeBase)
+	addr, n16 := allocator.allocNode16()
+	testFn(&artNode{kind: typeNode16, addr: addr}, &n16.nodeBase)
+	addr, n48 := allocator.allocNode48()
+	testFn(&artNode{kind: typeNode48, addr: addr}, &n48.nodeBase)
+	addr, n256 := allocator.allocNode256()
+	testFn(&artNode{kind: typeNode256, addr: addr}, &n256.nodeBase)
+}
+
+func TestOrderedChild(t *testing.T) {
+	// the keys of node4 and node16 should be kept ordered when adding new children.
+	var allocator artAllocator
+	allocator.init()
+
+	testFn := func(fn func() *artNode, getKeys func(*artNode) []byte, cap int) {
+		// add children in ascend order
+		an := fn()
+		for i := 0; i < cap; i++ {
+			addr, _ := allocator.allocNode4()
+			an.addChild(&allocator, byte(i), false, artNode{kind: typeNode4, addr: addr})
+
+			node := an.asNode(&allocator)
+			require.Equal(t, node.nodeNum, uint8(i+1))
+			keys := getKeys(an)
+			for j := 0; j <= i; j++ {
+				require.Equal(t, byte(j), keys[j])
+			}
+		}
+
+		// add children in descend order
+		an = fn()
+		for i := 0; i < cap; i++ {
+			addr, _ := allocator.allocNode4()
+			an.addChild(&allocator, byte(255-i), false, artNode{kind: typeNode4, addr: addr})
+
+			node := an.asNode(&allocator)
+			require.Equal(t, node.nodeNum, uint8(i+1))
+			keys := getKeys(an)
+			for j := 0; j <= i; j++ {
+				require.Equal(t, byte(255-i+j), keys[j])
+			}
+		}
+	}
+
+	testFn(func() *artNode {
+		addr, _ := allocator.allocNode4()
+		return &artNode{kind: typeNode4, addr: addr}
+	}, func(an *artNode) []byte {
+		return an.asNode4(&allocator).keys[:]
+	}, 4)
+
+	testFn(func() *artNode {
+		addr, _ := allocator.allocNode16()
+		return &artNode{kind: typeNode16, addr: addr}
+	}, func(an *artNode) []byte {
+		return an.asNode16(&allocator).keys[:]
+	}, 16)
+}
+
+func TestNextPrevPresentIdx(t *testing.T) {
+	// the present bitmap is used to find the next/prev present child index in node48 and node256.
+	var allocator artAllocator
+	allocator.init()
+
+	testFn := func(an *artNode, node interface {
+		nextPresentIdx(int) int
+		prevPresentIdx(int) int
+	}) {
+		require.Equal(t, node.nextPresentIdx(0), node256cap)
+		require.Equal(t, node.prevPresentIdx(0), inplaceIndex)
+
+		mid := 128
+		addr, _ := allocator.allocNode4()
+		an.addChild(&allocator, byte(mid), false, artNode{kind: typeNode4, addr: addr})
+		for i := 0; i < node256cap; i++ {
+			expectedNext := mid
+			expectedPrev := inplaceIndex
+			if i > mid {
+				expectedNext = node256cap
+			}
+			if i >= mid {
+				expectedPrev = mid
+			}
+			require.Equal(t, node.nextPresentIdx(i), expectedNext)
+			require.Equal(t, node.prevPresentIdx(i), expectedPrev)
+		}
+	}
+
+	addr, n48 := allocator.allocNode48()
+	testFn(&artNode{kind: typeNode48, addr: addr}, n48)
+	addr, n256 := allocator.allocNode48()
+	testFn(&artNode{kind: typeNode48, addr: addr}, n256)
+}
```

Comment:
Is there unit test for art node operations like expansion or recursive insert? It's better to cover these basic pathes.

Reply:
Added the expansion node and leaf tests.